bi-att-flow-dev/basic/model_nqa.py

```diff
@@ -183,114 +203,99 @@ def _build_forward(self):
-                concat_flat_yp2 = tf.nn.softmax(concat_flat_logits2)
-                na_prob2 = tf.squeeze(tf.slice(concat_flat_yp2, [0, 0], [-1, 1]), [1])  # [N]
-                flat_yp2 = tf.slice(concat_flat_yp2, [0, 1], [-1, -1])
-
-                self.concat_logits = concat_flat_logits
-                self.concat_logits2 = concat_flat_logits2
-                self.na_prob = na_prob * na_prob2
-
-            yp = tf.reshape(flat_yp, [-1, M, JX])
-            yp2 = tf.reshape(flat_yp2, [-1, M, JX])
-            wyp = tf.nn.sigmoid(logits2)
-
-            self.tensor_dict['g1'] = g1
-            self.tensor_dict['g2'] = g2
-
-            self.logits = flat_logits
-            self.logits2 = flat_logits2
-            self.yp = yp
-            self.yp2 = yp2
-            self.wyp = wyp
+
+            # (fw_g1, _), (my_fw_final_state, _) = bidirectional_dynamic_rnn(second_cell_fw, second_cell_bw, g0, x_len, dtype='float', scope='g1')  # [N, M, JX, 2d]
+            # g1 = tf.concat(axis=3, values=[fw_g1, bw_g1]) # g1 seems to be M in paper
+            # FIXME: Concatenate my_fw_final_state and my_bw_final_state
+
+            flat_g0 = flatten(g0, 2)  # [-1, J, d]
+            flat_x_len = None if x_len is None else tf.cast(flatten(x_len, 0), 'int64')
+
+            _, my_fw_final_state = tf.nn.dynamic_rnn(
+                cell = second_cell_fw,
+                inputs = flat_g0,
+                sequence_length = flat_x_len,
+                initial_state=None,
+                dtype='float',
+                parallel_iterations=None,
+                swap_memory=False,
+                time_major=False,
+                scope='my_fw_final_state'
+            ) # FIXME: Replace with bi-rnn
+
+            # print(tf.shape(my_fw_final_state)) # Tensor("model_0/main/Shape_14:0", shape=(3,), dtype=int32, device=/device:GPU:0)
+            # print(my_fw_final_state) # LSTMStateTuple(c=<tf.Tensor 'model_0/main/g1/fw/fw/while/Exit_2:0' shape=(?, 100) dtype=float32>, h=<tf.Tensor 'model_0/main/g1/fw/fw/while/Exit_3:0' shape=(?, 100) dtype=float32>)
+
+            # For now go with my_fw_final_state only
+
+            # Decoder embedding
+
+            # Embedding decoder/matrix
+
+            tgt_vocab_size = VW # hparam # FIXME: Obtain embeddings differently?
+            print(tgt_vocab_size)
+            tgt_embedding_size = dw # hparam
+            print(tgt_embedding_size)
+
+            embedding_decoder = word_emb_mat
+            # tf.Variable(
+            #     tf.constant(0.0, shape=[tgt_vocab_size, tgt_embedding_size]), trainable=False, name="embedding_decoder"
+            # )
+
+            # Look up embedding
+            decoder_emb_inp = tf.nn.embedding_lookup(embedding_decoder, self.decoder_inputs) # [batch_size, max words, embedding_size]
+            print(decoder_emb_inp)
+
+            def decode(helper, scope, reuse=None, maximum_iterations=None):
```

Comment:
i would suggest that instead of using a function(decode function here), put the entire decoder code in a separate variable scope